### PR TITLE
Check return values on update operations.

### PR DESCRIPTION
**Pull request by uwiger**
- put(), write() and delete() wrapped, checking for non-ok returns
- Option {on_write_error, verbose | warning | error | fatal}
  added to the rocksdb_opts user property. The corresponding function
  mnesia_lib:Op(Fmt, Args) will be called if an error return is spotted.

Addressing issue #12 - It's possible that writes have failed silently on AWS systems, but we're not logging anything about them.

No special tests written yet. Here are some shell commands, after changing the `write_result(...)` function to treat every return value as an error to be reported (i.e. lines 1592-593 commented out):

```erlang
Erlang/OTP 20 [erts-9.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V9.3  (abort with ^G)
(ulf@ulf-mbxp)1> mnesia:create_schema([node()]).
ok
(ulf@ulf-mbxp)2> mnesia:start().
ok
(ulf@ulf-mbxp)3> mnesia_rocksdb:register().
{ok,rocksdb_copies}
...
(ulf@ulf-mbxp)7> mnesia:create_table(e,[{rocksdb_copies,[node()]},{user_properties,[{rocksdb_opts,[{on_write_error,error}]}]}]).

=ERROR REPORT==== 24-Sep-2019::15:34:22 ===
Mnesia('ulf@ulf-mbxp'): ** ERROR ** FAILED rocksdb:delete(#Ref<0.3376793163.2625765402.93249>, <<1,12,
                                                                                                 185,218,
                                                                                                 111,86,
                                                                                                 80,8>>, []) -> ok
{atomic,ok}
(ulf@ulf-mbxp)8> mnesia:dirty_write({e,a,1}).

=ERROR REPORT==== 24-Sep-2019::15:34:48 ===
Mnesia('ulf@ulf-mbxp'): ** ERROR ** FAILED rocksdb:put(#Ref<0.3376793163.2625765402.93249>, <<12,176,
                                                                                              128,8>>, <<131,104,
                                                                                                         3,100,0,
                                                                                                         1,101,
                                                                                                         106,97,1>>, []) -> ok
ok
(ulf@ulf-mbxp)9> mnesia:create_table(w,[{rocksdb_copies,[node()]},{user_properties,[{rocksdb_opts,[{on_write_error,warning}]}]}]).

=WARNING REPORT==== 24-Sep-2019::15:35:23 ===
Mnesia('ulf@ulf-mbxp'): ** WARNING ** FAILED rocksdb:delete(#Ref<0.3376793163.2625765402.93251>, <<1,12,
                                                                                                   185,218,
                                                                                                   111,86,
                                                                                                   80,8>>, []) -> ok
{atomic,ok}
(ulf@ulf-mbxp)10> mnesia:dirty_write({w,a,1}).                                                                                     
=WARNING REPORT==== 24-Sep-2019::15:35:38 ===
Mnesia('ulf@ulf-mbxp'): ** WARNING ** FAILED rocksdb:put(#Ref<0.3376793163.2625765402.93251>, <<12,176,
                                                                                                128,8>>, <<131,104,
                                                                                                           3,100,0,
                                                                                                           1,119,
                                                                                                           106,97,1>>, []) -> ok
ok
(ulf@ulf-mbxp)11> mnesia:create_table(f,[{rocksdb_copies,[node()]},{user_properties,[{rocksdb_opts,[{on_write_error,fatal}]}]}]).  

=ERROR REPORT==== 24-Sep-2019::15:35:59 ===
Mnesia('ulf@ulf-mbxp'): ** ERROR ** (core dumped to file: "/home/uwiger/ae/mnesia_rocksdb/MnesiaCore.ulf@ulf-mbxp_1569_332159_506630")
 ** FATAL ** FAILED rocksdb:delete(#Ref<0.3376793163.2625765402.93253>, <<1,12,
                                                                          185,218,
                                                                          111,86,
                                                                          80,8>>, []) -> ok

=ERROR REPORT==== 24-Sep-2019::15:36:09 ===
** Generic server 'mnesia_ext_proc_e-_tab' terminating 
** Last message in was {'EXIT',<0.87.0>,killed}
** When Server state == {st,'mnesia_ext_icache_e-_tab',
                            #Ref<0.3376793163.2625765402.93249>,
                            rocksdb_copies,e,set,0,false,error}
** Reason for termination == 
** killed

=ERROR REPORT==== 24-Sep-2019::15:36:09 ===
** Generic server 'mnesia_ext_proc_w-_tab' terminating 
** Last message in was {'EXIT',<0.87.0>,killed}
** When Server state == {st,'mnesia_ext_icache_w-_tab',
                            #Ref<0.3376793163.2625765402.93251>,
                            rocksdb_copies,w,set,0,false,warning}
** Reason for termination == 
** killed

=INFO REPORT==== 24-Sep-2019::15:36:09 ===
    application: mnesia
    exited: killed
    type: temporary
```

The meaning of the `on_write_error` options are as follows:
* `debug` causes a printout only if the Mnesia debug level is set to `debug` or `trace`
* `warning` always causes a printout, but doesn't change the return value
* `error` always prints, and raises a (noisy) exception for the operation
* `fatal` prints, forces a mnesia core dump and termination of mnesia

```erlang
Erlang/OTP 20 [erts-9.3] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V9.3  (abort with ^G)
(ulf@ulf-mbxp)1> mnesia:create_schema([node()]).
ok
(ulf@ulf-mbxp)2> mnesia:start().
ok
(ulf@ulf-mbxp)3> mnesia_rocksdb:register().
{ok,rocksdb_copies}
(ulf@ulf-mbxp)4> mnesia:create_table(d,[{rocksdb_copies,[node()]},{user_properties,[{rocksdb_opts,[{on_write_error,debug}]}]}]).
{atomic,ok}
(ulf@ulf-mbxp)5> mnesia:dirty_write({d,a,1}).  
ok
(ulf@ulf-mbxp)6> mnesia:set_debug_level(debug).
none
(ulf@ulf-mbxp)7> mnesia:dirty_write({d,a,1}).  
Mnesia('ulf@ulf-mbxp'): FAILED rocksdb:put(#Ref<0.3142186912.214827034.101843>, <<12,176,
                                                                                  128,8>>, <<131,104,
                                                                                             3,100,0,
                                                                                             1,100,
                                                                                             106,97,1>>, []) -> ok
```

### CODE REVIEW

File: src/mnesia_rocksdb.erl [hunk 1]
```diff
@@ -182,6 +182,7 @@
 	    , type
 	    , size_warnings			% integer()
 	    , maintain_size			% boolean()
+            , on_write_error = verbose          :: debug | verbose | warning | error | fatal
```
Comment (tolbrino) [inline src/mnesia_rocksdb.erl]:
```suggestion
          , on_write_error = verbose          :: debug | verbose | warning | error | fatal
```

Comment (tolbrino) [reply src/mnesia_rocksdb.erl]: The possible values should be documented. You already described some of them in the PR.

File: src/mnesia_rocksdb.erl [hunk 2]
```diff
@@ -338,21 +339,36 @@ prefixes(_) ->
...
+                  if T==set; T==ordered_set; T==bag ->
+                          P;
+                     true ->
+                          mnesia:abort({combine_error,
+                                        Tab,
+                                        [Id, {type,T}]})
+                  end;
+             ({user_properties, UPs} = P) ->
+                  RdbOpts = proplists:get_value(rocksdb_opts, UPs, []),
+                  case proplists:get_value(on_write_error, RdbOpts) of
```
Comment (tolbrino) [inline src/mnesia_rocksdb.erl]: By using the same default here as in the type specification, you could get rid of a case statement here.

File: src/mnesia_rocksdb.erl [hunk 3]
```diff
@@ -1555,6 +1571,55 @@ keypat_pfx({HeadPat,_Gs,_}, KeyPos) when is_tuple(HeadPat) ->
...
+    write_result(?rocksdb:delete(Ref, K, Opts), delete, [Ref, K, Opts], St).
+
+write_result(ok, _, _, _) ->
+    ok;
+write_result(Res, Op, Args, #st{on_write_error = Rpt}) ->
+    RptOp = rpt_op(Rpt),
+    mnesia_lib:RptOp("FAILED rocksdb:~p(" ++ rpt_fmt(Args) ++ ") -> ~p~n",
+                   [Op | Args] ++ [Res]),
+    if Rpt == fatal; Rpt == error ->
+            badarg;
```
Comment (tolbrino) [inline src/mnesia_rocksdb.erl]: A generic `badarg` to cover all error cases will make handling errors in clients quite hard. Is there no way to pass the error further up the call chain?

Comment (uwiger) [reply src/mnesia_rocksdb.erl]: The only places where the callback module can trigger a client-side error are `validate_key/6` and `validate_record/6`. The actual update operations happen after the commit point, so the alternatives are basically:
* fail silently
* log/report, then fail silently
* log/report, then crash - `badarg` - causing termination of the calling process (internal mnesia error)
* dump core

Comment (tolbrino) [reply src/mnesia_rocksdb.erl]: Ok, so in the client we can only retry. If that fails too we have to bail out entirely. Or we build a lager tracer which inspects log messages from Mnesia and can be queried given a transaction identifier.

**Pull request merged**